### ISSUE TITLE
Optional[...] = None in Pydantic model froze at compiling FSM in ```outlines.generate.json(model, schema)```

### ISSUE THREAD

**Issue by whatswrongwithyourmitochondria** [dottxt-ai/outlines#1266]
### Describe the issue as clearly as possible:

I got a problem while creating a generator with ```outlines.generate.json(model, schema)``` and as a schema using Pydantic Model with the fields like ```name: Optional[...] = Field(default=None,...)```. It froze at compiling FSM, basically, at crawl() computing map for the state. 

In the example provided I put 1/4 of the Pydantic Model I have, overall schema is massive. 
I was able to fix that part getting rid of default=None in Optional[...] fields.
After removing it, the generator was created in seconds. 

### Steps/code to reproduce the bug:

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline

from pydantic import BaseModel, Field, field_validator
from pydantic.json_schema import models_json_schema

import outlines
from outlines import models, generate, types
from typing import Literal, Optional, List
from enum import Enum

import pandas as pd
import json
import torch

from huggingface_hub import login
from google.colab import userdata
login(token=userdata.get('HF_TOKEN'))

torch.random.manual_seed(0)

model_name = "meta-llama/Llama-3.1-8B-Instruct"
quantization_config = BitsAndBytesConfig(load_in_8bit=True)
model = AutoModelForCausalLM.from_pretrained(
                    model_name,
                    device_map="auto",
                    torch_dtype=torch.float16,
                    quantization_config=quantization_config
                    )

tokenizer = AutoTokenizer.from_pretrained(model_name)
llm = models.Transformers(model, tokenizer)

class MorphologyType(str, Enum):
    solid_polypoid = "solid_polypoid"
    solid_semi_annular = "solid_semi_annular"
    mucinous = "mucinous"
    other = "OTHER"
    missing = "missing"

class PeritonealReflectionRelation(str, Enum):
    below = "below"
    above = "above"
    crossing = "crossing"
    missing = "missing"

class TStageType(str, Enum):
    t12 = "T1_2"
    t3ab = "T3ab"
    t3cd = "T3cd"
    t4 = "T4"
    missing = "missing"

class SphincterInvasionType(str, Enum):
    no = "no"
    internal_only = "yes, internal sphincter only"
    internal_intersphincteric= "yes, internal sphincter + intersphincteric plane"
    internal_external = "yes, internal + external sphincter"
    unknown = "unknown"

class InvasionPart(str, Enum):
    upper = "upper"
    middle = "middle"
    distal = "distal"
    other = "OTHER"
    na = "not applicable"


class LocalTumorStatus(BaseModel):
    morphology: MorphologyType = Field(
        default=MorphologyType.missing,
        alias='Morphology: [solid_polypoid/solid_semi_annular/mucinous]'
    )
    circumferential_tumor_involvement_start: Optional[int] = Field(
        default=None,
        alias='Circumferential tumor involvement from [ ] o’clock'
    )
    circumferential_tumor_involvement_end: Optional[int] = Field(
        default=None,
        alias='Circumferential tumor involvement: to [ ] o’clock'
    )
    distance_from_anorectal_junction_cm: Optional[float] = Field(
        default=None,
        alias='Distance from anorectal junction to lower pole of tumor in cm'
    )
    relation_to_anterior_peritoneal_reflection: PeritonealReflectionRelation = Field(
        default=PeritonealReflectionRelation.missing,
        alias='Relation to anterior peritoneal reflection: [below/above/crossing]'
    )
    tumor_length_cm: Optional[float] = Field(
        default=None,
        alias='Tumor length in cm'
    )
    t_stage: TStageType = Field(
        default=TStageType.missing,
        alias='T_stage: [T1_2/T3ab/T3cd/T4]'
    )

    growth_into: Optional[str] = Field(
        default=None,
        alias='If T4, based on growth into: [ ]'
    )

    sphincter_invasion: SphincterInvasionType = Field(
        default=SphincterInvasionType.unknown,
        alias='Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]'
    )
    lowest_part_of_sphincter_invasion: InvasionPart = Field(
        default=InvasionPart.na,
        alias='If yes, lowest part of invasion [upper/middle/distal]'
    )

report_processor = generate.json(llm, LocalTumorStatus)
```

### Expected result:

```shell
getting a generator
```

### Error message:

```shell
the execution 'route':
Executing (3m 8s) <cell line: 1> ➔ wrapper() ➔ json() ➔ wrapper() ➔ regex() ➔ __init__() ➔ from_regex() ➔ from_regex() ➔ wrapper() ➔ cached_create_states_mapping() ➔ create_states_mapping() ➔ to_fsm() ➔ <genexpr> ➔ to_fsm() ➔ to_fsm() ➔ __mul__() ➔ times() ➔ crawl()
```

### Outlines/Python version information:

Outlines version
```0.1.3```
Python version
```Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]```
```



### Context for the issue:

The code is executed in Colab on NVIDIA L4 GPU.

**Comment by whatswrongwithyourmitochondria** [comment]
Unfortunately, without a default value equaling None the model hallucinates and output random numbers in the corresponding fields.

**Comment by whatswrongwithyourmitochondria** [comment]
I checked the regex obtained from schema without the defaults and with the defaults. For some reason with the defaults regex gets much bigger and with repetitions:

regex without defaults:
<details>
  <summary>Click to expand/collapse</summary>
  \{([ ]?"Morphology:\ \[solid_polypoid/solid_semi_annular/mucinous\]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?[ ]?"Circumferential\ tumor\ involvement\ from\ \[\ \]\ o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,[ ]?"Circumferential\ tumor\ involvement:\ to\ \[\ \]\ o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,[ ]?"Distance\ from\ anorectal\ junction\ to\ lower\ pole\ of\ tumor\ in\ cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,([ ]?"Relation\ to\ anterior\ peritoneal\ reflection:\ \[below/above/crossing\]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?[ ]?"Tumor\ length\ in\ cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,([ ]?"T_stage:\ \[T1_2/T3ab/T3cd/T4\]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing")[ ]?,)?[ ]?"If\ T4,\ based\ on\ growth\ into:\ \[\ \]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null)([ ]?,[ ]?"Sphincter\ invasion:\ \[no/yes,\ internal\ sphincter\ only/yes,\ internal\ sphincter\ \+\ intersphincteric\ plane/yes,\ internal\ \+\ external\ sphincter\]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If\ yes,\ lowest\ part\ of\ invasion\ \[upper/middle/distal\]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?[ ]?\}
  
  </details>
  
  regex with defaults:
  <details>
  <summary>Click to expand/collapse</summary>
  \{([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")([ ]?,[ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null))?([ ]?,[ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null))?([ ]?,[ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing"))?([ ]?,[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?[ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)([ ]?,[ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null))?([ ]?,[ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing"))?([ ]?,[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?[ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)([ ]?,[ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing"))?([ ]?,[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?[ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)([ ]?,[ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing"))?([ ]?,[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?[ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")([ ]?,[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null))?([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?[ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)([ ]?,[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing"))?([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?([ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?[ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing")([ ]?,[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null))?([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?([ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing")[ ]?,)?[ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null)([ ]?,[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown"))?([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?([ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing")[ ]?,)?([ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null)[ ]?,)?[ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown")([ ]?,[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?|([ ]?"Morphology: [solid_polypoid/solid_semi_annular/mucinous]"[ ]?:[ ]?("solid_polypoid"|"solid_semi_annular"|"mucinous"|"OTHER"|"missing")[ ]?,)?([ ]?"Circumferential tumor involvement from [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Circumferential tumor involvement: to [ ] o’clock"[ ]?:[ ]?((-)?(0|[1-9][0-9]*)|null)[ ]?,)?([ ]?"Distance from anorectal junction to lower pole of tumor in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"Relation to anterior peritoneal reflection: [below/above/crossing]"[ ]?:[ ]?("below"|"above"|"crossing"|"missing")[ ]?,)?([ ]?"Tumor length in cm"[ ]?:[ ]?(((-)?(0|[1-9][0-9]*))(\.[0-9]+)?([eE][+-][0-9]+)?|null)[ ]?,)?([ ]?"T_stage: [T1_2/T3ab/T3cd/T4]"[ ]?:[ ]?("T1_2"|"T3ab"|"T3cd"|"T4"|"missing")[ ]?,)?([ ]?"If T4, based on growth into: [ ]"[ ]?:[ ]?("([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])*"|null)[ ]?,)?([ ]?"Sphincter invasion: [no/yes, internal sphincter only/yes, internal sphincter + intersphincteric plane/yes, internal + external sphincter]"[ ]?:[ ]?("no"|"yes,\ internal\ sphincter\ only"|"yes,\ internal\ sphincter\ \+\ intersphincteric\ plane"|"yes,\ internal\ \+\ external\ sphincter"|"unknown")[ ]?,)?[ ]?"If yes, lowest part of invasion [upper/middle/distal]"[ ]?:[ ]?("upper"|"middle"|"distal"|"OTHER"|"not\ applicable"))?[ ]?\}
  
  </details>

**Comment by Touzen** [comment]
I have a similar problem. I create my model dynamically:
```
keys_and_their_types = {
    tag_name: (Optional[conlist(str, max_length=5, min_length=0)], None)
    for tag_name in ner_tag_names
}

```
and `len(ner_tag_names)` is typically around 8.

Each time I kill the process with ctrl+C it seems to be stuck on line 1000 in `fsm.py`:
<details><summary>Stack trace</summary>
<p>

```
Traceback (most recent call last):
  File "prompt_ner.py", line 50, in <module>
    main(args.conll, args.model)
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "prompt_ner.py", line 37, in main
    generator = outlines.generate.json(model, pydantic_model)
  File "/usr/lib64/python3.13/functools.py", line 934, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/outlines/generate/json.py", line 52, in json
    generator = regex(model, regex_str, sampler)
  File "/usr/lib64/python3.13/functools.py", line 934, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/outlines/generate/regex.py", line 38, in regex
    logits_processor = RegexLogitsProcessor(regex_str, tokenizer=model.tokenizer)
  File "venv/lib/python3.13/site-packages/outlines/processors/structured.py", line 151, in __init__
    guide = RegexGuide.from_regex(regex_string, tokenizer)
  File "venv/lib/python3.13/site-packages/outlines/fsm/guide.py", line 92, in from_regex
    return super().from_regex(
           ~~~~~~~~~~~~~~~~~~^
        regex_string,
        ^^^^^^^^^^^^^
    ...<2 lines>...
        **kwargs,
        ^^^^^^^^^
    )
    ^
  File "venv/lib64/python3.13/site-packages/outlines_core/fsm/guide.py", line 212, in from_regex
    ) = _create_states_mapping(
        ~~~~~~~~~~~~~~~~~~~~~~^
        regex_string,
        ^^^^^^^^^^^^^
    ...<2 lines>...
        frozen_tokens=frozen_tokens,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "venv/lib/python3.13/site-packages/outlines/fsm/guide.py", line 76, in cached_create_states_mapping
    return uncached_create_states_mapping(regex_string, tokenizer, *args, **kwargs)
  File "venv/lib64/python3.13/site-packages/outlines_core/fsm/guide.py", line 140, in create_states_mapping
    regex_fsm = regex_parser(regex_string).to_fsm()
  File "venv/lib/python3.13/site-packages/interegular/patterns.py", line 453, in to_fsm
    return FSM.union(*(o.to_fsm(alphabet, prefix_postfix, flags) for o in self.options))
           ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/interegular/patterns.py", line 453, in <genexpr>
    return FSM.union(*(o.to_fsm(alphabet, prefix_postfix, flags) for o in self.options))
                       ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/interegular/patterns.py", line 396, in to_fsm
    current.append(part.to_fsm(alphabet, (0, 0), flags))
                   ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/interegular/patterns.py", line 288, in to_fsm
    unit = self.base.to_fsm(alphabet, (0, 0), flags=flags)
  File "venv/lib/python3.13/site-packages/interegular/patterns.py", line 453, in to_fsm
    return FSM.union(*(o.to_fsm(alphabet, prefix_postfix, flags) for o in self.options))
           ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "venv/lib/python3.13/site-packages/interegular/fsm.py", line 457, in union
    return parallel(fsms, any)
  File "venv/lib/python3.13/site-packages/interegular/fsm.py", line 943, in parallel
    return crawl(alphabet, initial, final, follow)
  File "venv/lib/python3.13/site-packages/interegular/fsm.py", line 1000, in crawl
    j = states.index(next)

```
</p>
</details> 

I also need to use optional lists to avoid hallucinations. I can see how having optional fields would make the search space larger, but it would be really nice to have some form of support for this behaviour.

**Comment by whatswrongwithyourmitochondria** [comment]
Hi, Touzen! I can only suggest having a look at your schema in the regex form:
```
from outlines_core.fsm.json_schema import build_regex_from_schema
import json as pyjson

reg_def = build_regex_from_schema(your_schema)
print(reg_def)
```

Btw, do you need your fields to be **required**? Or you use Optional keeping in mind that those fields are **not required**?

In my case I also added Ellipsis to indicate that those Optional fields are **required**, and for some reason I was able to create a generator reasonably fast.

**Comment by jbbqqf** [comment]
Hi! Triaging — this report uses `outlines.generate.json(model, schema)` with an `Optional[...] = Field(default=None)` Pydantic field hanging at FSM compilation (through `interegular`).

Evidence on current `main` (`6ae2135`):
- `generate.json` was removed in v1 — migration guide https://github.com/dottxt-ai/outlines/blob/6ae21356fd0dda00afd0ef0e76752904a22fc0aa/docs/guide/migration.md#L19, release note https://github.com/dottxt-ai/outlines/blob/6ae21356fd0dda00afd0ef0e76752904a22fc0aa/outlines/release_note.md#L40. JSON output is now produced by passing a Pydantic class / TypedDict / JsonSchema as the output type to a `Generator`.
- The FSM-compile-via-`interegular` path that hangs in the report has been displaced; v1.x JSON generation goes through the backends in https://github.com/dottxt-ai/outlines/blob/6ae21356fd0dda00afd0ef0e76752904a22fc0aa/outlines/backends (e.g. `outlines_core`, `xgrammar`, `llguidance`) rather than the v0 `interegular.crawl()`-driven FSM build.

In other words, the specific hot loop (`crawl()` building the state map for an `Optional` field with `default=None`) isn't on the v1 code path. If anyone still sees `Optional[...] = None` cause a hang with v1.x `Generator(...)`, that would warrant a fresh issue with a backend-tagged reproducer. Otherwise this is a candidate for closing as superseded.

---
*Disclosure: I drafted this comment with help from Claude Code while triaging stale issues on `outlines` v1.x. The code/doc references above were verified manually against `main` (`6ae2135`, current latest 1.3.0).*

**Comment by ErenAta16** [comment]
The actual mechanism here is in `outlines_core`'s JSON-schema-to-regex compiler (`parse_properties` in `src/json_schema/parsing.rs`), not in `outlines` itself. It has two code paths: if your schema has *at least one* required property, it emits one linear pattern where every other field is just wrapped as an independent `(field,)?` -- cheap to compile. But if **no property is required** -- which is exactly what happens when every field in a Pydantic model has `default=...`, since Pydantic then leaves the schema's `"required"` list empty -- the compiler instead builds *N* separate full alternative patterns (one per field), each carrying a nested optional-prefix of every field before it. That's why your "regex with defaults" blob is so much bigger than the no-defaults version. When that gets turned into an FSM, the union/crawl step has to explore a state space that's roughly the product of all those branches, and it scales combinatorially with the number of all-optional fields -- with ~9 of them, it just never finishes.

The practical fix isn't "don't use Optional," it's "make sure at least one property is required in the schema." You can keep the *value* nullable while still marking the *key* required, e.g. with Pydantic: `field: Optional[int] = Field(...)` (no default) instead of `Field(default=None)` -- this still lets the model emit `"field": null`, but keeps at least one entry in `required`, which routes you through the cheap linear branch instead of the exponential one. That should get you the "compiles in seconds" behavior you saw when you removed the defaults, without fully giving up nullable fields.

Longer term this is worth a real fix upstream in `outlines_core` -- the all-optional case shouldn't need N full alternative branches; it could use the same "each field is an independent `(x)?` in a fixed order" strategy the required-case uses (JSON key order isn't semantically required to be preserved, and outlines already doesn't guarantee arbitrary key ordering).
